Update configuration to always download HTML/JS/XML Bitstreams (no inline display)

tdonohue opened the pull request:
## Description
This PR updates our default settings for `webui.content_disposition_format` to include HTML, JavaScript, XML and RDF by default.  

The changes in this PR are the equivalent of these settings in `dspace.cfg`.  These settings will work in any DSpace 7.6 or 7.6.1 installation to provide the same fixes as this PR.
```
webui.content_disposition_format = text/html
webui.content_disposition_format = text/javascript
webui.content_disposition_format = text/xml
webui.content_disposition_format = rdf
```

This ensures that bitstreams of these formats are always downloaded and never opened inline by a user's browser.  Opening these formats inline _may_ cause the browser to execute any JavaScript contained within them. (Execution of embedded JavaScript may not always occur for all sites, but it is possible.)

**New Wildcard Option**

This PR also provides sites the ability to force *all* formats to download only.  This means that no files will ever be opened inline in a user's browser.  To force all formats to download only, use a wildcard/asterisk:
```
webui.content_disposition_format = *
```

**Add JavaScript to Format Registry**

This PR also adds JavaScript to the list of default formats in the Bitstream Format Registry to ensure it is a recognized format (by default).

## Instructions for Reviewers
* Deploy this PR
* Upload one or more HTML or JS or XML or RDF bitstreams to a new or existing Item
* Attempt to download those bitstreams.  Ensure do *not* open in your browser window.  Instead, they should download directly to your machine.
* Test the new wildcard option.
    * Set `webui.content_disposition_format = *` in your local.cfg or dspace.cfg
    * Ensure all file formats will now **only download**.  
    * For example, your browser should no longer attempt to open PDFs or image files in the browser window.  These will all be downloaded.

## Related work

For more on `webui.content_disposition_format` see #8891

tdonohue committed:
Avoid inline display of HTML/JS bitstreams.  Add JS to list of known formats so that it can be recognized by DSpace.

mwoodiupui reviewed (commented):
Looks right but seems to be failing integration tests.

tdonohue commented:
@mwoodiupui : Thanks for the note!  It appears we have an IT which is still *hardcoding* a specific count of BitstreamFormats (sigh).  Since this PR added a single new BitstreamFormat those counts all failed.  So, it's really a flawed test & not a sign of an issue in this PR.

Nonetheless, I'll get the test fixed (hopefully later today) so that it succeeds.  (If I can figure out a way to easily remove the hardcoded numbers, I'll do so...but not sure how much effort that'd be)

tdonohue committed:
Fix failing IT by increasing number of formats by one

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/BitstreamFormatRestRepositoryIT.java:
```diff
@@ -56,7 +57,16 @@ public class BitstreamFormatRestRepositoryIT extends AbstractControllerIntegrati
     private BitstreamFormatConverter bitstreamFormatConverter;
 
-    private final int DEFAULT_AMOUNT_FORMATS = 85;
+    // Default number of formats in the registry (filled out in @Before method below)
```
I know that it is boring fix this kind of stuff often but it is more.appropriate in my opinion an hardcoded check here. If we rely on our API to perform the check we loss the guarantee about which registries are loaded by default. This would hide unwanted changes.

tdonohue replied:
@abollini : I have to admit I get frustrated by hardcoded numbers like this, but I'll revert it as it's really unrelated to this PR.

tdonohue replied:
Reverted the change to make the count dynamic.  Now I just updated the hardcoded value: [8d81c14](https://github.com/DSpace/DSpace/pull/9638/commits/8d81c144bba435a5b515ae6e54cd65582f343eda)

abollini reviewed (changes requested):
I'm ok with this PR except for the IT fix that I would recommend to fix keeping the number of types hardcoded

tdonohue commented:
@mwoodiupui : I fixed the failing tests by ensure the `BitstreamFormatRestRepositoryIT` no longer hardcodes the number of BitstreamFormats.  The updated code now loads that count from the BitstreamFormatService before tests run, ensuring it will update dynamically for the future.

mwoodiupui commented:
Perhaps, **separate from this issue**, we should have a discussion about why the test should depend on a magic number that has to be changed when unrelated improvements are made.

mwoodiupui reviewed (approved):
+1 by inspection

Xib3rR4dAr commented:
XML files (text/xml) also allow XSS.

tdonohue committed:
Add XML to the list of formats to always download

tdonohue committed:
Add a test to prove the default settings are to always download these formats

tdonohue commented:
I've updated this PR today to include the following:
* Added XML (text/xml) to the list of formats to always download
* Added automated tests to `BitstreamRestControllerIT` to verify that HTML, JS & XML are all downloaded by default.  (This is to simply protect us and other sites from removing these configs in the future)

tdonohue committed:
For additional security, ensure "unknown" formats are always downloaded. Update test to prove behavior.

mwoodiupui commented:
I hate to drag this out further, but if we list `text/xml` then we probably also need `application/xml`.  https://stackoverflow.com/q/4832357/2916377

tdonohue committed:
To avoid misconfiguration, hardcode HTML, XML, RDF, JS to download only. Add a new wildcard setting to allow sites to force all files to download only.

tdonohue commented:
@abollini and @mwoodiupui :  I've made some changes to this PR today which I hope are improvements, so I'd appreciate a re-review.  See the updated PR description for all the details...but here are the highlights:
* I've hardcoded some file formats into Java code to avoid misconfiguration.  While I don't tend to like hardcoded settings, I think it may be best to do so in this situation.  For example, I cannot think of a use case where these settings should NOT be in place.
* I've added a new wildcard (asterisk: `*`) option to disable inline opening for **all files**.
* I've added more automated tests to verify everything is working as it should.

tdonohue commented:
@mwoodiupui : We don't need to include `application/xml` in this PR as that's not a format in our DSpace Format Registry: https://github.com/DSpace/DSpace/blob/main/dspace/config/registries/bitstream-formats.xml

Any file format that is NOT in that registry will be returned as `application/octet-stream`.  Browsers seem to **always** download `application/octet-stream`.  But, I've also added some automated tests & minor code cleanup in my last commit to force anything that comes back as `application/octet-stream` to use `content-disposition: attachment`

tdonohue commented:
Final note for today:  Because everything that comes back as `application/octet-stream` now uses `content-disposition: attachment`, we *could* now **remove** `text/javascript` from the registry (i.e. revert those changes from this PR).  I don't believe it is needed anymore.  But, I'll wait to do anything like that until I have further feedback on the directions of this PR.

(We could also wait until this week's Developer's Meeting to discuss all this in greater detail together.)

pnbecker reviewed (approved):
I did a code review. It looks good to me. I have one comment: while I understand why we have a hard-coded restriction on html, javascript and xml, we do not have a way to override this. There are probably repositories that want to be able to show html after properly reviewing it or by allowing very trusted users to submit anything. I think it would be less than one hour to introduce a new configuration property that allows to switch off this hardcoded list.

Seeing the timeline for DSpace 8 and the importance of this patch I'm +1.

abollini reviewed (approved):
I'm ok with the current PR. Unfortunately, in DSpace 7+ we loose the possibility to serve properly html content that wourl require relative url to work to download css and additional content in the same item. So force the download is the safer and easier solution at this time. We could reconsider it when we have a full functional html support

kshepherd reviewed (approved):
+1 tested successfully and the code looks good to me.

Yes, it is a 'hard-coded' fix, but sets the safest defaults, and I think any implementer who is willing to take the risk (perhaps they have extra validation of uploaded html/xml/js) of serving this content inline could reasonably be expected to apply a patch that reverts this change for DSpace 8.x
